share/doc/wake/logging.md

```diff
@@ -0,0 +1,64 @@
+You can also make your own stream using `mkLogLevel`.
+You can use this to customize the color you want it to be shown as,
+or (more commonly) to grab output from a `Job` for passing to another `Job`,
+consuming within your wake program, or redirecting outside of wake using shell redirect.
+
+Unless overridden with `--fd:2=...` the *only* thing that ever appears on stderr coming out of wake is `logError`,
+and it always appears there.
+
+For stdout, the following table should be read top to bottom considering the command line arguments you've passed to wake.
+If you match a line in the table, stop!
+This is what will appear on the console as you watch wake execute, and what is sent to wake's stdout:
+
+|Flag               | logDebug | logInfo | logEcho | logReport | logWarning | logError | logNever | mkLogLevel "foo" Green|
+|-------------------|----------|---------|---------|-----------|------------|----------|----------|----------------------
+|Color on Console   |  Blue    |  Dim    |  default| Magenta   |  Yellow    |   Red    |          |   Green         |          
+|--fd:1="foo"       |          |         |         |           |            |          |          |         x       |
+|--fd:1="foo,info"  |          |    x    |         |           |            |          |          |         x       |
+|--no-execute       |          |         |         |           |            |    x     |          |                 |
+|--debug            |      x   |    x    |    x    |     x     |      x     |    x     |          |                 |
+|--verbose          |          |    x    |    x    |     x     |      x     |    x     |          |                 |
+|--quiet            |          |         |         |           |            |    x     |          |                 |
+|--no-tty           |          |         |    x    |     x     |      x     |    x     |          |                 |
+|(default)          |          |         |         |     x     |      x     |    x     |          |                 |
+
+`println` and `print` go to `logReport`. You can make print statements that go to other levels with `printlnLevel` or `printLevel`.
+These statements are *not* saved to the wake database, no matter which `stream` they are sent to.
+
+Of course one can always manually redirect or tee stdin/stdout to files within the command executed by the `Job`, outside
+of these mechanisms.
+
+## How Logging works on Wake Database Commands
+
+Confusingly, the command line arguments in the database alias with those when executing, but they have different behavior.
+
+Running `wake <database command> --verbose` will show ALL of stdout and stdin from all `Job`s, regardless of what wake
+stream they were directed to. Running a database command without `--verbose` will show NO stdout or stdin from any `Job`.
+
+Recall that results of `print` or `println` are not stored in the database, so will never be revealed with any combination
+of flags.
```

Comment:
It doesn't matter whether you can/can't do this, the point is that people should not expect to see it. It's not super intuitive what the database is *storing*. If I run `wake -v -o <some output>` it is reasonable user expectation to see the same results as I would have seen by running `wake -v <some command that generated some output>`.

I think all this is under development and changing, and the plan would be to print it out with the database commands "more like the user would have seen it" by running wake -v, and then I think it *will* become even more confusing that the printlns don't show up in that view.

Reply:
I can see that it may be unintuitive what the database is storing, but I think that's completely orthogonal to logging, which is what this document is about.

> If I run wake -v -o <some output> it is reasonable user expectation to see the same results as I would have seen by running wake -v <some command that generated some output>.

I'm not entirely sure that I agree with that, but I think it depends on what someone thinks `-o` is supposed to do. Once again, I'm going to claim that this should be explained in a separate document about the various commands that query job information from the database, not the document about logging, since logging doesn't affect the database commands.

Reply:
Maybe. I guess this is a question of "is this a document for someone writing wake code, or someone runnign wake code, or both"?

If it's for the latter category, the experience is "I run a wake command. A bunch of stuff spews to my screen. I see if I run `wake --last` I can see some of that stuff." I'd like to give more intuition on what that stuff is, and why some of it doesnt' show up.

Let me ask a different related question. Are there other things in wake aside from this whole logging mechanism and print/println that make things show up on the user's console? I'd like to make that in the scope of this document (the first part), then I'd like to consider what to say about it, if anything, in the second part.

Like, does returning a Fail result automatically put something on the console (not sure this sentence even makes sense)? Is that reflected in the database?

Reply:
> Maybe. I guess this is a question of "is this a document for someone writing wake code, or someone runnign wake code, or both"?

I assumed the former, since none of the log level stuff is really relevant to someone only running `wake`, since as someone only running `wake` rules built by others, you're really just at the whims of whatever LogLevels were chosen by the person who authored the rules.

> Let me ask a different related question. Are there other things in wake aside from this whole logging mechanism and print/println that make things show up on the user's console? I'd like to make that in the scope of this document (the first part), then I'd like to consider what to say about it, if anything, in the second part.
>
> Like, does returning a Fail result automatically put something on the console (not sure this sentence even makes sense)? Is that reflected in the database?

You brought up the only other thing that I can think of, which is the final returned value is printed to screen. This is true regardless of whatever the return value is, i.e. `Pass` or `Unit` or anything just get printed at the end. I believe they go to stdout, which is one of the reasons why you _have_ to map your logging streams to fd:3 or some other file descriptor or else the final return value gets included into stdout.

Reply:
Sorry, didn't answer the final question, but no, this is not reflected in the database because it's not a Job. The wake database holds Job information, and Jobs are only created with `runJob`.

Nothing else is a Job.

Reply:
>  since none of the log level stuff is really relevant to someone only running wake

I mean, they see things getting printed out in different colors, some of which show up when they do or do not specify different flags, so I do think that is very relevant to the users.

Reply:
(I guess I am saying I think I am mixing the two users/reader types indiscriminately in this document, which is sloppy, I can think about more clearly dividing it in more of a 'so you are writing wake code' vs 'so you're running wake commands').

Reply:
ok, I tried to break this out more clearly into things that are about jobs running and "other stuff". let me know what you think.